File: frontend/src/api/tasks.ts
import { api } from "./client";

export interface Task {
  id: number;
  title: string;
  description: string | null;
  due_date: string | null;
  status: "open" | "done" | "cancelled";
  contact_id: number;
  assignee_id: number | null;
  created_at: string;
  completed_at: string | null;
}

export async function listTasks(filters?: {
  status?: string;
  contact_id?: number;
  assignee_id?: number;
}): Promise<Task[]> {
  const { data } = await api.get("/api/tasks", { params: filters });
  return data;
}

export async function createTask(payload: {
  title: string;
  description?: string | null;
  due_date?: string | null;
  contact_id: number;
  assignee_id?: number | null;
}): Promise<Task> {
  const { data } = await api.post("/api/tasks", payload);
  return data;
}

export async function updateTask(id: number, payload: Partial<Task>): Promise<Task> {
  const { data } = await api.patch(`/api/tasks/${id}`, payload);
  return data;
}

export async function deleteTask(id: number): Promise<void> {
  await api.delete(`/api/tasks/${id}`);
}

File: backend/app/routers/tasks.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.deps import get_current_user
from app.models import Contact, Task, User
from app.schemas import Message, TaskIn, TaskOut, TaskUpdate

router = APIRouter(prefix="/api/tasks", tags=["tasks"])


@router.get("", response_model=list[TaskOut])
async def list_tasks(
    status: str | None = None,
    contact_id: int | None = None,
    assignee_id: int | None = None,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> list[Task]:
    stmt = select(Task).where(Task.organization_id == user.organization_id)
    if status:
        stmt = stmt.where(Task.status == status)
    if contact_id is not None:
        stmt = stmt.where(Task.contact_id == contact_id)
    if assignee_id is not None:
        stmt = stmt.where(Task.assignee_id == assignee_id)
    stmt = stmt.order_by(Task.due_date.is_(None), Task.due_date.asc(), Task.created_at.desc())
    res = await db.execute(stmt)
    return list(res.scalars().all())


@router.post("", response_model=TaskOut)
async def create_task(
    payload: TaskIn,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> Task:
    contact = await db.get(Contact, payload.contact_id)
    if contact is None or contact.organization_id != user.organization_id:
        raise HTTPException(status_code=404, detail="Contact not found")
    task = Task(
        title=payload.title,
        description=payload.description,
        due_date=payload.due_date,
        contact_id=payload.contact_id,
        assignee_id=payload.assignee_id or user.id,
        organization_id=user.organization_id,
    )
    db.add(task)
    await db.commit()
    await db.refresh(task)
    return task


@router.patch("/{task_id}", response_model=TaskOut)
async def update_task(
    task_id: int,
    payload: TaskUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> Task:
    task = await db.get(Task, task_id)
    if task is None or task.organization_id != user.organization_id:
        raise HTTPException(status_code=404, detail="Task not found")
    data = payload.model_dump(exclude_unset=True)
    if "status" in data and data["status"] == "done" and task.status != "done":
        task.completed_at = datetime.now(timezone.utc)
    if "status" in data and data["status"] != "done":
        task.completed_at = None
    for field, value in data.items():
        setattr(task, field, value)
    await db.commit()
    await db.refresh(task)
    return task


@router.delete("/{task_id}", response_model=Message)
async def delete_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
) -> Message:
    task = await db.get(Task, task_id)
    if task is None or task.organization_id != user.organization_id:
        raise HTTPException(status_code=404, detail="Task not found")
    await db.delete(task)
    await db.commit()
    return Message(detail="ok")
